src/macaron/repo_finder/provenance_extractor.py

```diff
@@ -232,34 +181,56 @@ def _extract_from_witness_provenance(payload: InTotoV01Payload) -> tuple[str, st
             repo = _json_extract(entry, ["attestation", "projecturl"], str)
 
     if not commit or not repo:
-        logger.debug("Could not extract repo and commit from provenance.")
-        return "", ""
+        raise ProvenanceExtractionException("Could not extract repo and commit from provenance.")
 
     return repo, commit
 
 
+def _extract_commit_from_digest_set(digest_set: dict[str, JsonType], valid_algorithms: list[str]) -> str:
+    """Extract the commit from the passed DigestSet.
+
+    The DigestSet is an in-toto object that maps algorithm types to commit hashes (digests).
+    """
+    # TODO decide on a preference for which algorithm to accept.
+    if len(digest_set.keys()) > 1:
+        logger.debug("DigestSet contains multiple algorithms: %s", digest_set.keys())
+
+    for key in digest_set:
+        if key in valid_algorithms:
+            value = digest_set.get(key)
+            if isinstance(value, str):
+                return value
+    raise ProvenanceExtractionException("No valid digest in digest set.")
+
+
+def _clean_spdx(uri: str) -> str:
+    """Clean the passed SPDX URI and return the normalised URL it represents.
+
+    A SPDX URI has the form: git+https://example.com@refs/heads/main
+    """
+    url, _, _ = uri.lstrip("git+").rpartition("@")
+    return url
+
+
 T = TypeVar("T", bound=JsonType)
 
 
-def _json_extract(entry: dict[str, JsonType], keys: list[str], type_: type[T]) -> T | None:
+def _json_extract(entry: dict[str, JsonType], keys: list[str], type_: type[T]) -> T:
```

Comment:
I am sorry for saying this after you have changed this function to raise the ProvenanceExtraction instead :sweat_smile: . 
For me I was find with this private method returning None for error. It's supposed to be a generic function that as long as `entry` is a JSON-liked object. 
I can see some issues with raising `ProvenanceExtractionException` here:
- Because this is supposed to be a generic function, raising a context specific exception `ProvenanceExtractionException` is not a good idea. 
- The exceptions raised here are bubbled up to the caller function (for example at this [place](https://github.com/oracle/macaron/blob/56e680e36bc87170109972ce7f114b71abd69b33/src/macaron/repo_finder/provenance_extractor.py#L72)), which to me is not very explicit.
I do have some suggestions:
- If we want to keep the exception here, I think we can rename it to something more generic (let's called it `JsonExtractionException` for example). But it also means that we have to explicitly catch this exception whenever we use it.
- We can make this function return None on error and check for None on usage without having to catch any errors.
I am fine with both suggestions so please let me know what you think.

Reply:
I agree, although utilising exceptions for the JSON extraction allows for cleaner code in the provenance extractor functions, in my opinion. My compromise is using suggestion 1 above, but the catching of the JSON extraction exception occurs in the public function of the file `extract_repo_and_commit_from_provenance`, which itself still only raises `ProvenanceExtractionException`.

Reply:
Yep, that sounds good to me. If I understand your comment correctly, you were mentioning about catching and re-raising the exceptions from this `_json_extract` method within the public function:
```python
try:
    list_index = _json_extract(predicate, ["recipe", "definedInMaterial"], int) 
except JsonExtractException as error:
    raise ProvenanceExtractionException(...) from error
```